### ISSUE TITLE
Semantic tokens not entire applied on buffer

### ISSUE THREAD

**Issue by ericdallo** [emacs-lsp/lsp-mode#3154]
### Thank you for the bug report

- [X] I am using the latest version of `lsp-mode` related packages.
- [X] I checked [FAQ](https://emacs-lsp.github.io/lsp-mode/page/faq/) and [Troubleshooting](https://emacs-lsp.github.io/lsp-mode/page/troubleshooting/) sections
- [X] You may also try reproduce the issue using clean environment using the following command
```bash
emacs -q -l lsp-start-plain.el
```
where `lsp-start-plain.el` can be downloaded [here](https://github.com/emacs-lsp/lsp-mode/blob/master/scripts/lsp-start-plain.el).

### Bug description

Sometimes when opening a buffer the semantic tokens are applied only for the current visible region (probably range semantic tokens) but if scrolling the buffer up or down, I don't see tokens for those other regions unless I `M-x` `lsp` again, even so other regions would not contain the tokens.

print:
![image](https://user-images.githubusercontent.com/7820865/137595904-4f57247c-b1ea-4e1f-a7af-157658aedb7b.png)
You can see how just part of the buffer has the green tokens

I tested on vscode with clojure-lsp and it works fine, so it seems a issue on lsp-mode.

### Steps to reproduce

You can clone https://github.com/clojure-lsp/clojure-lsp/ and open `server.clj` buffer, it's possible to see in other buffers as well.

### Expected behavior

All tokens correctly applied for the entire buffer.

### Which Language Server did you use?

[clojure-lsp](https://github.com/clojure-lsp/clojure-lsp/)

### OS

Linux

### Anything else?

@sebastiansturm could you help understand why that happens? I was thinking that was some issue on server but after implementing semantic tokens support on Calva(VScode extension to code clojure) [here](https://github.com/BetterThanTomorrow/calva/pull/1336), everything worked perfectly there.
I think this is the only issue I see with semantic tokens right now, but it's quite annoying as it happens a lot :/

**Comment by ericdallo** [comment]
Thank you for taking a look @sebastiansturm!

**Comment by ericdallo** [comment]
I'm also having issues with `cider` package that is the REPL package that almost every clojure user uses it, it seems cider apply a lot of font lock and I can't disable that accurately, but I think this is other issue :)

**Comment by ericdallo** [comment]
Good point, I'm not sure as well but I think we can test those options, I'm not familiar with the lsp-mode semantic tokens code but I'm with the server side code, even so LMK if you need any help with testing any of those options

**Comment by sebastiansturm** [comment]
had another look at clojure-mode and noticed it merely pushes a new element onto `font-lock-extend-region-functions` (of which I had been unaware); will try if applying the same to our request regions yields good results.

**Comment by sebastiansturm** [comment]
didn't seem like such a good fit either because font-lock requests chunk boundaries that don't coincide with the window boundaries I'm using for ranged requests. As a compromise, I'm now overshooting by several jit-lock-chunk-sizes in both directions, so the newly tokenized region is likely to include any extended region close to the current viewport, while still being potentially much smaller than the entire buffer. Though as we're later requesting the full set on idle anyway, I'm not sure if this actually is a useful optimization or just unneeded overhead.
Anyway, I'll push the change to my repo at sebastiansturm/lsp-mode in a few minutes, would appreciate if you could try that out. Have tested it a few times on clojure buffers and one rust buffer, but haven't used it in anger yet

**Comment by ericdallo** [comment]
Great @sebastiansturm! Thanks and sure, I can test it!

**Comment by ericdallo** [comment]
Gentle ping @sebastiansturm :)

**Comment by sebastiansturm** [comment]
sorry for not following up sooner. My intention was to combine this with a fix to issue #3155 which during my cursory testing seemed to work, but I'd like to see that verified. Though on the other hand, no point in delaying the PR any further so I'll open one and outline what I think should best be re-evaluated by someone other than me

**Comment by ericdallo** [comment]
thanks @sebastiansturm I just pinged you because your branch seems to really improve the range tokens for me, feel free to wait if you think it makes sense, both works to me :)
